phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,535 @@
+                }
+                LOGGER.info("Loaded Indexer coprocessor on data table " + tableName);
+            } else {
+                LOGGER.info("Indexer coprocessor on data table " + tableName + " is already loaded");
+            }
+        }
+        if (!dryRun) {
+            admin.modifyTable(dataTableDescBuilder.build());
+        }
+    }
+
+    private void modifyIndexTable(Admin admin, HashSet<String> indexes)
+            throws IOException {
+        for (String indexName : indexes) {
+            TableDescriptorBuilder indexTableDescBuilder = TableDescriptorBuilder
+                    .newBuilder(admin.getDescriptor(TableName.valueOf(indexName)));
+            if (upgrade) {
+                if (!admin.getDescriptor(TableName.valueOf(indexName))
+                        .hasCoprocessor(GlobalIndexChecker.class.getName())) {
+                    if (!dryRun) {
+                        indexTableDescBuilder.addCoprocessor(GlobalIndexChecker.class.getName(),
+                                null, QueryServicesOptions.DEFAULT_COPROCESSOR_PRIORITY, prop);
+                    }
+                    LOGGER.info("Loaded GlobalIndexChecker coprocessor on index table " + indexName);
+                } else {
+                    LOGGER.info("GlobalIndexChecker coprocessor on index table " + indexName + " is already loaded");
+                }
+            } else {
+                if (admin.getDescriptor(TableName.valueOf(indexName))
+                        .hasCoprocessor(GlobalIndexChecker.class.getName())) {
+                    if (!dryRun) {
+                        indexTableDescBuilder.removeCoprocessor(GlobalIndexChecker.class.getName());
+                    }
+                    LOGGER.info("Unloaded GlobalIndexChecker coprocessor on index table "+indexName);
+                } else {
+                    LOGGER.info("GlobalIndexChecker coprocessor on index table " + indexName + " is already unloaded");
+                }
+            }
+            if (!dryRun) {
+                admin.modifyTable(indexTableDescBuilder.build());
+            }
+        }
+    }
+
+
+    private void rebuildIndexes(String dataTable, Configuration conf) {
+        String schema = SchemaUtil.getSchemaNameFromFullName(dataTable);
+        String table = SchemaUtil.getTableNameFromFullName(dataTable);
+        for(String index: rebuildMap.get(dataTable)) {
+            String outFile = "/tmp/index_rebuild_"+index+"_" + UUID.randomUUID().toString();
+            String indexName = SchemaUtil.getTableNameFromFullName(index);
+            conf.set(QueryServices.TRANSACTIONS_ENABLED, Boolean.TRUE.toString());
+            indexingTool.setConf(conf);
+            String [] args = {"-s", schema, "-dt", table, "-it", indexName, "-direct", "-runfg", "-op", outFile};
+            try {
+                LOGGER.info("Rebuilding index "+ index);
+                if(!dryRun) {
+                    indexingTool.run(args);
```

Comment:
Since this is running in the foreground, and some indexes take hours to rebuild, this can take a very long time. Should we be running in the background, or multi-threading? (That would of course be more complicated because you'd need callbacks or polling)

Reply:
@swaroopak - Since @kadirozde 's recent change to make unknown == UNVERIFIED, should the rebuild step be optional? I can see some operators wanting to separate the coprocessor upgrade process from the rebuild process.

Reply:
Will go with optionally rebuilding as sync or async :)